src/codegen/expression.rs

```diff
@@ -2397,30 +2423,52 @@ pub fn emit_function_call(
                 );
 
                 // If the first element of returns is Void, we can discard the returns
-                if !ftype.returns.is_empty() && returns[0] != Type::Void {
-                    let mut returns = Vec::new();
-                    let mut res = Vec::new();
+                if !dest_func.returns.is_empty() && returns[0] != Type::Void {
+                    let mut returns;
 
-                    for ret in &*ftype.returns {
-                        let id = pt::Identifier {
-                            loc: ret.loc,
-                            name: ret.name_as_str().to_owned(),
-                        };
-                        let temp_pos = vartab.temp(&id, &ret.ty);
-                        res.push(temp_pos);
-                        returns.push(Expression::Variable(id.loc, ret.ty.clone(), temp_pos));
-                    }
+                    if dest_func.selector.is_some() {
+                        // borsh decode
+                        let mut types = Vec::new();
 
-                    cfg.add(
-                        vartab,
-                        Instr::AbiDecode {
-                            res,
-                            selector: None,
-                            exception_block: None,
-                            tys: (*ftype.returns).clone(),
-                            data: Expression::ReturnData(*loc),
-                        },
-                    );
+                        for ret in &*dest_func.returns {
+                            types.push(ret.ty.clone());
+                        }
+
+                        let encoder = create_encoder(ns);
+
+                        returns = encoder.abi_decode(
+                            loc,
+                            &Expression::ReturnData(*loc),
+                            &types,
+                            ns,
+                            vartab,
+                            cfg,
+                        );
+                    } else {
+                        returns = Vec::new();
+                        let mut res = Vec::new();
+
+                        for ret in &*dest_func.returns {
+                            let id = pt::Identifier {
+                                loc: ret.loc,
+                                name: ret.name_as_str().to_owned(),
+                            };
+                            let temp_pos = vartab.temp(&id, &ret.ty);
+                            res.push(temp_pos);
+                            returns.push(Expression::Variable(id.loc, ret.ty.clone(), temp_pos));
+                        }
+
+                        cfg.add(
+                            vartab,
+                            Instr::AbiDecode {
```

Comment:
Why aren't we using Borsh here?

Reply:
This is the case when we're calling solidity functions, so we're still using eth abi encoding here for now.